apis/python/src/tiledbsoma/io/_caching_reader.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import annotations
+
+import ctypes
+import io
+import sys
+import threading
+from collections import OrderedDict
+from typing import Any, Union, cast
+
+import attrs
+import pyarrow as pa
+from typing_extensions import Buffer, TypeAlias
+
+# typeguard and cython memoryview incompat in older versions
+if sys.version_info >= (3, 12):
+    WritableBuffer: TypeAlias = Union[memoryview, Buffer]
+else:
+    WritableBuffer: TypeAlias = Any
+
+
+@attrs.define
+class CacheStats:
+    block_idx: int
+    hit: int = 0
+    miss: int = 0
+
+
+class CachingReader:
+    """Buffering reader which maintains a LRU cache of previously read data, wrapping
+    and presenting a file-like interface. Typical use is similar to ``io.BufferedReader``:
+
+        f = CachingReader(open("a file", mode="rb"), memory_budget=64*1024**2, cache_block_size=1024**2)
+
+    The LRU cache is user-defined by:
+        * memory budget - total cache size in bytes
+        * block_size - the unit of cached data. Also the read size
+
+    Instances present as a file-like object, and all public methods conform to the normal Python
+    ``io`` package semantics. Most, but not all methods required for io.RawIOBase (read-only) are
+    implemented.
+
+    Class assumes that the underlying file is not changing concurrent with read.
+
+    Primary use case is wrapping a VFS FileBuffer object, to improve performance of read
+    operations by doing fewer, larger reads.
+
+    The implemtation stores cached data as PyArrow arrays, primarily to benefit from the
+    easy, minimal-copy concat and slice operations upon underlying buffers.
+    """
+
+    def __init__(
+        self,
+        file: Any,  # file-like object. Unfortunately, Python lacks a good typing signature for this concept.
```

Comment:
Optional: We could use either a protocol or `clib.SOMAVFSFilebuf` for typing here, although I think the comment is probably sufficient.

Reply:
I had several thoughts:
* I wanted to leave the option open for other file-like objects to use this in the future.  Not sure that is very compelling though :-)
* this module doesn't depend on any behavior specific to `clib.SOMAVFSFilebuf`
* typing it as `clib.SOMAVFSFilebuf` is the same as `Any`, as that type is not defined as far as mypy knows (if you do a `reveal_type` on anything from `clib`, they are all `Any` from a static typing perspective)